### PR TITLE
KAFKA-14966 - Extract reusable functionality from OffsetFetcher

### PR DESCRIPTION
**Pull request by lianetm**
Initial refactoring for extracting reusable logic from OffsetFetcher into OffsetFetcherUtils. No changes in logic, just moving classes, functions or pieces of logic into OffsetFetcherUtils. 

All the functionality moved from OffsetFetcher to the OffsetFetcherUtils is already covered by tests in OffsetFetcherTest and FetcherTest. There were no individual tests for the extracted functions, so no tests were migrated.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcherUtils.java [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Utility functions for fetching offsets, validating and resetting positions.
+ */
+class OffsetFetcherUtils {
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcherUtils.java:46]: I wonder if it is too hard to make this a static class.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcherUtils.java [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    private final ConsumerMetadata metadata;
+    private final SubscriptionState subscriptionState;
+    private final Time time;
+    private final ApiVersions apiVersions;
+    private final Logger log;
+
+    private final AtomicReference<RuntimeException> cachedOffsetForLeaderException = new AtomicReference<>();
+    private final AtomicInteger metadataUpdateVersion = new AtomicInteger(-1);
+
+    OffsetFetcherUtils(LogContext logContext,
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcherUtils.java:56]: nit: Please align the constructor parameters.

**Review by kirktrue — CHANGES_REQUESTED** [review]
Yep, it's a straightforward refactor, like you said 😄 

Really only have the one nit about parameter alignment.

Another reason to have `OffsetFetcherUtils` remain as an instance is that it maintains state (e.g. `metadataUpdateVersion`) which wouldn't work as easily if it were just a grab bag of static methods.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcherUtils.java [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        this.apiVersions = apiVersions;
+    }
+
+    /**
+     * Callback for the response of the list offset call.
+     * @param listOffsetsResponse The response from the server.
+     * @return {@link OffsetFetcherUtils.ListOffsetResult} extracted from the response, containing the fetched offsets
+     * and partitions to retry.
+     */
+    OffsetFetcherUtils.ListOffsetResult handleListOffsetResponse(ListOffsetsResponse listOffsetsResponse) {
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcherUtils.java:74]: This is actually a request handler.  I might make sense to create a separated class for it (maybe nested inside one of the fetcher class), WDYT? Taking this away might make it easier to "static-ify" the class.  I've been doing that a lot for the refactored consumers.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcherUtils.java (hunk 1 above)
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcherUtils.java:46]: The thing here is that this class has shared stated among the functions, like @kirktrue pointed out in the comment below (including `subscriptionState`, `cachedOffsetForLeaderException`, `metadata`, etc.). Taking all that state out would remove most of the value of this class which is to reuse common logic. So, seen like that, WDYT? Should we maybe reconsider the Utils name if that's what is causing confusion?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcherUtils.java (hunk 3 above)
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcherUtils.java:74]: Definitely doable, but wonder if it's worth it in the case that we are not looking for turning this into a static class given that it maintains some state? The refactor is intentionally only extracting reusable code (not changing it), to keep the risk low. That being said, open to talk more about it, just want to make sure that any changes we introduce (other than extracting existing code) bring value that outweighs the risk

**Comment by philipnee** [comment]
LGTM! A couple of things:
1. Could you attach a Jira Number to the title: KAFKA-abcdefg Extracting reusable functions..
2. Could you verify if the tests are passing, verbally :)

**Comment by lianetm** [comment]
Thanks both for the comments. 

All client test pass (there were only some unrelated broker-side test failures)

**Comment by lianetm** [comment]
Merged into AK on [PR-13815](https://github.com/apache/kafka/pull/13815)

**Pull request merged**